File: internal/commands/workflows.go
package commands

import (
	"fmt"
	"strings"

	"github.com/loki-bedlam/reposwarm-cli/internal/api"
	"github.com/loki-bedlam/reposwarm-cli/internal/output"
	"github.com/spf13/cobra"
)

func newWorkflowsCmd() *cobra.Command {
	cmd := &cobra.Command{
		Use:     "workflows",
		Aliases: []string{"wf"},
		Short:   "Manage Temporal workflows",
	}
	cmd.AddCommand(newWorkflowsListCmd())
	cmd.AddCommand(newWorkflowsStatusCmd())
	cmd.AddCommand(newWorkflowsTerminateCmd())
	return cmd
}

func newWorkflowsListCmd() *cobra.Command {
	var limit int

	cmd := &cobra.Command{
		Use:   "list",
		Short: "List recent workflows",
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := getClient()
			if err != nil {
				return err
			}

			var result api.WorkflowsResponse
			path := fmt.Sprintf("/workflows?pageSize=%d", limit)
			if err := client.Get(ctx(), path, &result); err != nil {
				return err
			}

			if flagJSON {
				return output.JSON(result.Executions)
			}

			fmt.Printf("\n  %s (%d workflows)\n\n", output.Bold("Workflows"), len(result.Executions))
			headers := []string{"Workflow ID", "Status", "Type", "Started"}
			var rows [][]string
			for _, w := range result.Executions {
				wfID := w.WorkflowID
				if len(wfID) > 50 {
					wfID = wfID[:47] + "..."
				}
				rows = append(rows, []string{
					wfID,
					output.StatusColor(w.Status),
					w.Type,
					w.StartTime,
				})
			}
			output.Table(headers, rows)
			fmt.Println()
			return nil
		},
	}

	cmd.Flags().IntVar(&limit, "limit", 25, "Max workflows to show")
	return cmd
}

func newWorkflowsStatusCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "status <workflow-id>",
		Short: "Show detailed workflow status",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			client, err := getClient()
			if err != nil {
				return err
			}

			var wf api.WorkflowExecution
			if err := client.Get(ctx(), "/workflows/"+args[0], &wf); err != nil {
				return err
			}

			if flagJSON {
				return output.JSON(wf)
			}

			fmt.Printf("\n  %s\n\n", output.Bold("Workflow Details"))
			fmt.Printf("  %s  %s\n", output.Dim("ID       "), wf.WorkflowID)
			fmt.Printf("  %s  %s\n", output.Dim("Run ID   "), wf.RunID)
			fmt.Printf("  %s  %s\n", output.Dim("Status   "), output.StatusColor(wf.Status))
			fmt.Printf("  %s  %s\n", output.Dim("Type     "), wf.Type)
			fmt.Printf("  %s  %s\n", output.Dim("Started  "), wf.StartTime)
			if wf.CloseTime != "" {
				fmt.Printf("  %s  %s\n", output.Dim("Closed   "), wf.CloseTime)
			}
			fmt.Println()
			return nil
		},
	}
}

func newWorkflowsTerminateCmd() *cobra.Command {
	var yes bool
	var reason string

	cmd := &cobra.Command{
		Use:   "terminate <workflow-id>",
		Short: "Terminate a running workflow",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if !yes {
				fmt.Printf("  Terminate workflow %s? [y/N] ", output.Bold(args[0]))
				var confirm string
				fmt.Scanln(&confirm)
				if strings.ToLower(confirm) != "y" {
					output.Infof("Cancelled")
					return nil
				}
			}

			client, err := getClient()
			if err != nil {
				return err
			}

			body := map[string]string{"reason": reason}
			var result any
			if err := client.Post(ctx(), "/workflows/"+args[0]+"/terminate", body, &result); err != nil {
				return err
			}

			if flagJSON {
				return output.JSON(map[string]any{"workflowId": args[0], "terminated": true})
			}
			output.Successf("Terminated workflow %s", args[0])
			return nil
		},
	}

	cmd.Flags().BoolVarP(&yes, "yes", "y", false, "Skip confirmation")
	cmd.Flags().StringVar(&reason, "reason", "Terminated via CLI", "Termination reason")
	return cmd
}

File: src/commands/workflows.ts
import { Command } from 'commander'
import chalk from 'chalk'
import { getGlobalOpts, output, error } from '../lib/output.js'

interface WorkflowExecution {
  workflowId: string
  runId: string
  status: string
  type: string
  startTime: string
  closeTime?: string
}

async function fetchWorkflows(temporalUrl: string, namespace: string, query?: string): Promise<WorkflowExecution[]> {
  const params = new URLSearchParams()
  if (query) params.set('query', query)

  const response = await fetch(
    `${temporalUrl}/api/v1/namespaces/${namespace}/workflows?${params}`,
    { headers: { 'Content-Type': 'application/json' } }
  )
  if (!response.ok) throw new Error(`Temporal API error: ${response.status}`)

  const data = await response.json() as { executions?: Array<{
    execution: { workflowId: string; runId: string }
    status: string
    type: { name: string }
    startTime: string
    closeTime?: string
  }> }
  return (data.executions || []).map(e => ({
    workflowId: e.execution.workflowId,
    runId: e.execution.runId,
    status: e.status,
    type: e.type.name,
    startTime: e.startTime,
    closeTime: e.closeTime
  }))
}

export const workflowsCommand = new Command('workflows')
  .description('Manage investigation workflows')

workflowsCommand
  .command('list')
  .description('List recent workflows')
  .option('--status <status>', 'Filter by status (Running, Completed, Failed)')
  .option('--limit <n>', 'Max results', '20')
  .option('--temporal-url <url>', 'Temporal HTTP API URL', 'http://temporal-alb-internal:8233')
  .option('--namespace <ns>', 'Temporal namespace', 'default')
  .action(async (opts, cmd) => {
    const global = getGlobalOpts(cmd)
    try {
      let query = ''
      if (opts.status) {
        const statusMap: Record<string, string> = {
          running: 'Running', completed: 'Completed', failed: 'Failed',
          terminated: 'Terminated', timedout: 'TimedOut'
        }
        const s = statusMap[opts.status.toLowerCase()] || opts.status
        query = `ExecutionStatus = "${s}"`
      }

      const workflows = await fetchWorkflows(opts.temporalUrl, opts.namespace, query)

      output(cmd, { count: workflows.length, workflows }, (data) => {
        if (data.count === 0) {
          console.log('No workflows found.')
          return
        }
        console.log(`\n${chalk.bold(data.count)} workflows:\n`)
        for (const wf of data.workflows) {
          const statusColor = wf.status === 'Running' ? chalk.blue
            : wf.status === 'Completed' ? chalk.green
            : wf.status === 'Failed' ? chalk.red
            : chalk.dim
          console.log(`  ${statusColor(wf.status.padEnd(12))} ${wf.workflowId}`)
          console.log(`  ${chalk.dim(`Started: ${wf.startTime}`)}`)
        }
        console.log()
      })
    } catch (err) {
      error(cmd, 'Failed to list workflows', err)
    }
  })

workflowsCommand
  .command('status <workflowId>')
  .description('Get workflow status')
  .option('--temporal-url <url>', 'Temporal HTTP API URL', 'http://temporal-alb-internal:8233')
  .option('--namespace <ns>', 'Temporal namespace', 'default')
  .action(async (workflowId, opts, cmd) => {
    const global = getGlobalOpts(cmd)
    try {
      const response = await fetch(
        `${opts.temporalUrl}/api/v1/namespaces/${opts.namespace}/workflows/${encodeURIComponent(workflowId)}`,
        { headers: { 'Content-Type': 'application/json' } }
      )
      if (!response.ok) throw new Error(`Temporal API error: ${response.status}`)
      const data = await response.json()

      output(cmd, data, (d) => {
        console.log(`\nWorkflow: ${chalk.bold(workflowId)}`)
        console.log(JSON.stringify(d, null, 2))
      })
    } catch (err) {
      error(cmd, `Failed to get workflow ${workflowId}`, err)
    }
  })

workflowsCommand
  .command('terminate <workflowId>')
  .description('Terminate a running workflow')
  .option('--reason <reason>', 'Termination reason', 'Terminated via CLI')
  .option('--temporal-url <url>', 'Temporal HTTP API URL', 'http://temporal-alb-internal:8233')
  .option('--namespace <ns>', 'Temporal namespace', 'default')
  .action(async (workflowId, opts, cmd) => {
    const global = getGlobalOpts(cmd)
    try {
      const response = await fetch(
        `${opts.temporalUrl}/api/v1/namespaces/${opts.namespace}/workflows/${encodeURIComponent(workflowId)}/terminate`,
        {
          method: 'POST',
          headers: { 'Content-Type': 'application/json' },
          body: JSON.stringify({ reason: opts.reason })
        }
      )
      if (!response.ok) throw new Error(`Temporal API error: ${response.status}`)

      output(cmd, { success: true, workflowId, reason: opts.reason }, () => {
        console.log(chalk.yellow(`Terminated workflow: ${workflowId}`))
      })
    } catch (err) {
      error(cmd, `Failed to terminate ${workflowId}`, err)
    }
  })
